### PR TITLE
[SPARK-5501][SPARK-5420][SQL] Write support for the data source API

### PR DESCRIPTION
**Pull request by yhuai**
This PR aims to support `INSERT INTO/OVERWRITE TABLE tableName` and `CREATE TABLE tableName AS SELECT` for the data source API (partitioned tables are not supported).

In this PR, I am also adding the support of `IF NOT EXISTS` for our ddl parser. The current semantic of `IF NOT EXISTS` is explained as follows. 
- For a `CREATE TEMPORARY TABLE` statement, it does not `IF NOT EXISTS` for now.
- For a `CREATE TABLE` statement (we are creating a metastore table), if there is an existing table having the same name ...
  - when `IF NOT EXISTS` clause is used, we will do nothing.
  - when `IF NOT EXISTS` clause is not used, the user will see an exception saying the table already exists.

TODOs:
- [x] CTAS support
- [x] Programmatic APIs
- [ ] Python API (another PR)
- [x] More unit tests
- [ ] Documents (another PR)

@marmbrus @liancheng @rxin

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala [hunk 1]
```diff
@@ -146,9 +146,13 @@ class SqlParser extends AbstractSparkSQLParser {
...
-    INSERT ~> OVERWRITE.? ~ (INTO ~> relation) ~ select ^^ {
-      case o ~ r ~ s => InsertIntoTable(r, Map.empty[String, Option[String]], s, o.isDefined)
-    }
+    ( INSERT ~> OVERWRITE ~> TABLE ~> relation ~ select ^^ {
+        case r ~ s => InsertIntoTable(r, Map.empty[String, Option[String]], s, true)
+      }
+    | INSERT ~> INTO ~> TABLE ~> relation ~ select ^^ {
+      case r ~ s => InsertIntoTable(r, Map.empty[String, Option[String]], s, false)
+      }
+    )
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala]: Can be simplified to this:

```scala
    INSERT ~> (OVERWRITE ^^^ true | INTO ^^^ false) ~ (TABLE ~> relation) ~ select ^^ {
      case o ~ r ~ s => InsertIntoTable(r, Map.empty[String, Option[String]], s, o)
    }
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -314,11 +314,17 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
   object DDLStrategy extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
-      case CreateTableUsing(tableName, userSpecifiedSchema, provider, true, options) =>
+      case CreateTableUsing(tableName, userSpecifiedSchema, provider, true, opts, allowExisting) =>
         ExecutedCommand(
-          CreateTempTableUsing(tableName, userSpecifiedSchema, provider, options)) :: Nil
+          CreateTempTableUsing(
+            tableName, userSpecifiedSchema, provider, opts, allowExisting)) :: Nil
+      case CreateTableUsing(_, _, _, false, _, _) =>
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Maybe `case c: CreateTableUsing if !c.temporary =>` for clarity and readability here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 3]
```diff
@@ -314,11 +314,17 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
+        sys.error("Tables created with SQLContext must be TEMPORARY. Use a HiveContext instead.")
 
-      case CreateTableUsing(tableName, userSpecifiedSchema, provider, false, options) =>
+      case CreateTableUsingAsSelect(tableName, provider, true, options, allowExisting, query) =>
+        ExecutedCommand(
+          CreateTempTableUsingAsSelect(tableName, provider, options, allowExisting, query)) :: Nil
+      case CreateTableUsingAsSelect(_, _, false, _, _, _) =>
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Same as above.

File: sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala [hunk 4]
```diff
@@ -28,10 +32,10 @@ private[sql] class DefaultSource extends RelationProvider with SchemaRelationPro
   override def createRelation(
       sqlContext: SQLContext,
       parameters: Map[String, String]): BaseRelation = {
-    val fileName = parameters.getOrElse("path", sys.error("Option 'path' not specified"))
+    val path = parameters.getOrElse("path", sys.error("Option 'path' not specified"))
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala]: I think we should define common data source options like `path` in a single place. But we can leave it to another PR.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 5]
```diff
@@ -145,6 +145,11 @@ abstract class PrunedFilteredScan extends BaseRelation {
...
  */
 @Experimental
-abstract class CatalystScan extends BaseRelation {
+trait CatalystScan extends BaseRelation {
   def buildScan(requiredColumns: Seq[Attribute], filters: Seq[Expression]): RDD[Row]
 }
+
+@DeveloperApi
+trait InsertableRelation extends BaseRelation {
+  def insertInto(data: DataFrame, overwrite: Boolean): Unit
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: The method name looks a bit weird when you put them together: `relation.insertInto(dataFrame)`, seems that we are inserting a relation into a data frame... Maybe just `insert`?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 6]
```diff
@@ -334,6 +335,71 @@ class SQLContext(@transient val sparkContext: SparkContext)
...
+
+  @Experimental
+  def load(
+    dataSourceName: String,
+    options: java.util.Map[String, String]): DataFrame = {
+    load(dataSourceName, options.toMap)
+  }
+
+  @Experimental
+  def loadAsTempTable(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: this seems redundant, since users can do

load(...).registerAsTempTable(...) ?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: btw the style convention is 4 space indent for function parameter defintion, i.e.

```scala
def loadAsTempTable(
    tableName: String,
    dataSourceName: String,
    options: Map[String, String]): Unit = {
  val cmd = ...
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 7]
```diff
@@ -334,6 +335,71 @@ class SQLContext(@transient val sparkContext: SparkContext)
     applySchema(rowRDD, appliedSchema)
   }
 
+  @Experimental
+  def load(
+    dataSourceName: String,
+    options: Map[String, String]): DataFrame = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: For the scala version I think it should just be `(String, String)*`.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 6 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: +1 to just chaining the calls.

File: sql/core/src/main/scala/org/apache/spark/sql/api.scala [hunk 8]
```diff
@@ -171,6 +171,27 @@ private[sql] trait DataFrameSpecificApi {
   def saveAsTable(tableName: String): Unit
 
   @Experimental
+  def saveAsTable(tableName: String, dataSourceName: String, options: Map[String, String]): Unit
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/api.scala]: When we add the short hand version that only takes a table name, we need to make sure that we set the `EXTERNAL` property correctly so that `DROP TABLE` will operate as users expect.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 7 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Then do we assume multiple values can be associated with a single key here?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: No we don't but I think there is any reason to force the syntactic overhead of saying `Map(...)` on the user.  If they put double keys in a `Map` they will also be silently discarded so I don't think we are regressing at all.

File: sql/core/src/main/scala/org/apache/spark/sql/api.scala [hunk 9]
```diff
@@ -171,6 +171,33 @@ private[sql] trait DataFrameSpecificApi {
...
+  def save(path: String): Unit
+
+  @Experimental
+  def save(path: String, overwrite: Boolean): Unit
+
+  @Experimental
+  def save(
+      dataSourceName: String,
+      overwrite: Boolean,
+      options: (String, String)*): Unit
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/api.scala]: i think this one might be ambiguous with the other save.

u'd want to declare something like
```scala
def save(
      dataSourceName: String,
      overwrite: Boolean,
      option: (String, String),
      options: (String, String)*): Unit
```

File: sql/core/src/main/scala/org/apache/spark/sql/api.scala [hunk 10]
```diff
@@ -171,6 +171,33 @@ private[sql] trait DataFrameSpecificApi {
...
+  def saveAsTable(
+      tableName: String,
+      dataSourceName: String,
+      options: java.util.Map[String, String]): Unit
+
+  @Experimental
+  def save(path: String): Unit
+
+  @Experimental
+  def save(path: String, overwrite: Boolean): Unit
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/api.scala]: as discussed offline, let's remove the overwrite version first

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 11]
```diff
@@ -618,8 +619,84 @@ class DataFrame protected[sql](
...
+  override def save(path: String, overwrite: Boolean): Unit = {
+    val dataSourceName = sqlContext.conf.defaultDataSourceName
+    save(dataSourceName, overwrite, ("path" -> path))
+  }
+
+  @Experimental
+  override def save(
+      dataSourceName: String,
+      overwrite: Boolean,
+      options: (String, String)*): Unit = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: here too. let's disambiguate it by having option: (String, String) followed by options: (String, String)

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 12]
```diff
@@ -618,8 +619,98 @@ class DataFrame protected[sql](
...
+
+  /**
+   * :: Experimental ::
+   * Creates a table from the the contents of this DataFrame based on a given data source and
+   * a set of options. This will fail if the table already exists.
+   *
+   * Note that this currently only works with DataFrames that are created from a HiveContext as
+   * there is no notion of a persisted catalog in a standard SQL context.  Instead you can write
+   * an RDD out to a parquet file, and then register that file as a table.  This "table" can then
+   * be the target of an `insertInto`.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: it'd be great to add inline code examples.
we can do this later as separate prs.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 13]
```diff
@@ -77,6 +77,14 @@ trait SchemaRelationProvider {
       schema: StructType): BaseRelation
 }
 
+trait CreateableRelation {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Change it to CreateableRelationProvider

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 14]
```diff
@@ -145,6 +153,11 @@ abstract class PrunedFilteredScan extends BaseRelation {
  * for experimentation.
  */
 @Experimental
-abstract class CatalystScan extends BaseRelation {
+trait CatalystScan extends BaseRelation {
   def buildScan(requiredColumns: Seq[Attribute], filters: Seq[Expression]): RDD[Row]
 }
+
+@DeveloperApi
+trait WritableRelation extends BaseRelation {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Change it to InsertableRelation

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 15]
```diff
@@ -145,6 +153,11 @@ abstract class PrunedFilteredScan extends BaseRelation {
...
+  def write(data: DataFrame, overwrite: Boolean): Unit
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: change it to insert

**Comment by rxin** [comment]
Thanks. I'm merging this in master.
